### PR TITLE
WIP: Add lint for missing Clone impls

### PR DESCRIPTION
**Pull request by Susurrus**
This is a WIP implementation for #58066. I've just implemented a lint for missing `Clone` implementations. For the other traits listed in #58066, I'll add separate commits for the missing ones once I've gotten some feedback here and am sure I'm on the right track.

Closes #58066.

### CODE REVIEW

**Comment by Susurrus** [comment]
Tests succeeded locally, so pushed the latest version here. I've only refactored the code and added both a lint for missing `Clone` implementations and a `missing_impls` lint group. Once I get a little feedback that this is the right path forward, I'll add the remaining trait lints (and add them to the group as well):

 * `Eq`
 * `PartialEq`
 * `Ord`
 * `PartialOrd`
 * `Hash`
 * `Display`
 * `Default`

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -529,136 +529,96 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingDoc {
...
-            _ => return,
-        };
-        if def.has_dtor(cx.tcx) {
-            return;
         }
-        let param_env = ty::ParamEnv::empty();
-        if ty.is_copy_modulo_regions(cx.tcx, param_env, item.span) {
-            return;
-        }
-        if param_env.can_type_implement_copy(cx.tcx, ty).is_ok() {
```
[the remark this answers is not in the document]
Comment (Susurrus) [reply src/librustc_lint/builtin.rs]: I don't think this warned you if the type couldn't be `Copy`, it just didn't trigger the lint. I'm wondering if this is really a problem in practice. I'd think that if you set this lint globally, you'd want it to trigger on all types and not exclude types that it can't be implemented for, so you'd have to whitelist them explicitly.

**Comment by Susurrus** [comment]
I started implementing `Hash` and found that my macro fails because there isn't a method to check if the given trait is implemented called `hash_trait`. This applies to: `PartialEq`, `Hash`, `Display`, `Default`. I'm looking at the `language_item_table!` macro call in `src/librustc/middle/lang_items.rs` and I'm wonder if it's a trivial fix to add these these in there, or if that actually breaks things by making new language items.

I also ran into problems implementing `Eq` (which I just updated this PR with), where if `PartialEq` is derived, the error doesn't show for `Eq` not being implemented. I'm uncertain why that is right now.

**Comment by Susurrus** [comment]
So looking at the [lang_items docs](https://doc.rust-lang.org/unstable-book/language-features/lang-items.html) I'm not seeing much clarity on what a `lang_item` actually is. As in, whether it'd be inappropriate here to define more. Only 3 traits need this treatment: `Default`, `Display`, and `Hash`. What're the rules around defining new lang items here? Does it break anything anywhere else? I'd think since they're in the stdlib, that this wouldn't be a big deal, but I probably don't have enough context.

**Comment by Susurrus** [comment]
I'm coming back to this because I actually went and ported this change to `rust-clippy`. But this implementation currently relies on these lang items being defined, so it fails there just as it fails here. So I need to rewrite if I can't add these 3 traits as lang items no matter where I put it. Any guidance here is appreciated.

**Comment by Susurrus** [comment]
Okay, I whipped up rust-lang/rust-clippy#3752, which has just Debug and Copy, which should be enough for us to finalize the work necessary here in rust.

I've changed this PR to mark those lints as deprecated and also added `Debug`, `Display`, and `Hash` as lang items so they can be easily checked the same way as all the other traits. This enables Clippy to easily implement all of these lints.

**Comment by oli-obk** [comment]
Ok, this is totally awesome. I talked with @Manishearth at the all hands about this and we have a way to be able to not affect the lang theam here: We add a new attribute to the compiler: `#[diagnostic = "debug_trait"]` and similar so we can stop checking for lang items and instead just check for diagnostic-specific items.

This will be a bit more involved. Essentially we need to

1. duplicate the `get_lang_items` query. Though I don't think there's any point in having a method per diagnostic item.
2. So I suggest we just make the query return a `FxHashMap<Symbol, NodeId>`.
3. We can then build a wrapper method on the `TyCtxt` that maps a `&str` to an `Option<NodeId>` for user convenience
4. Throw the new attribute on a bunch of types (basically go through https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/utils/paths.rs and just flag all the things in that list)
5. see if there are any rustc diagnostics that could benefit from this
6. replace all uses of `util/path.rs` with a use of the `&str` -> `Option<NodeId>` method
7. complain to @oli-obk that they didn't think this through and there's uses of `util/path.rs` that do something different
8. figure out if we need any of this in `EarlyLintPass`es and make them `LateLintPasses`
9. see 7.

**Comment by Centril** [comment]
(The lang team is involved here not because of lang items, but because new lints are being added... once @oli-obk feels the implementation here is up to snuff ping me and I'll FCP-merge...)

File: src/librustc_lint/builtin.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (scottmcm) [reply src/librustc_lint/builtin.rs]: I think this is super important.  I would never use a lint suggesting `Copy` if I had to suppress it any time I put a `Vec` or a `HashSet` inside my struct.

**Comment by oli-obk** [comment]
This PR does not add any new lints anymore. They are going to be implemented in clippy and the two rustc lints are going to get deprecated

**Comment by Centril** [comment]
OK... hmm... why are we deprecating the lints tho? I was happy for this to be done in rustc personally.

**Comment by oli-obk** [comment]
It's not a correctness thing, and they have to be enabled explicitly anyway. So users have to discover them. I think the discoverability of clippy lints is much higher than of allow-by-default rustc lints.

**Comment by Centril** [comment]
> I think the discoverability of clippy lints is much higher than of allow-by-default rustc lints.

This might be an issue of not having a page such as https://rust-lang.github.io/rust-clippy/master/index.html for the rustc lints; if we had that then rustc lints would become more discoverable?

**Comment by oli-obk** [comment]
Even so, if we don't consider them important enough to be enabled by default, why do we have them in the compiler? People looking for more static analysis usually go to clippy I presume?

**Comment by Centril** [comment]
Presumably we have them in the compiler because clippy wasn't a well known thing at that point in time? What other allow-by-default lints do we have? If we're going for deprecation on the basis of "allow by default should be in clippy" then a more holistic approach seems preferable?

**Comment by oli-obk** [comment]
They might be older than clippy XD

Here's a list of all allow-by-default lints in rustc that are not edition related to best of my knowledge

* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L32
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L38
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L111
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L117
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L239
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L245
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L257
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L294
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L312
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L318
* https://github.com/rust-lang/rust/blob/c67d4749a547a0c937e257ad9e0248075178ddc4/src/librustc/lint/builtin.rs#L350
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/builtin.rs#L103
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/builtin.rs#L224
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/builtin.rs#L316
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/builtin.rs#L539
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/builtin.rs#L603
* https://github.com/rust-lang/rust/blob/cb6fafbdf3b09da28b09fa7e87a24658d77dd02f/src/librustc_lint/unused.rs#L30
* https://github.com/rust-lang/rust/blob/cb6fafbdf3b09da28b09fa7e87a24658d77dd02f/src/librustc_lint/unused.rs#L459
* https://github.com/rust-lang/rust/blob/827a141466e7bb85eb3c030600878a2c606019e9/src/librustc_lint/types.rs#L43
*

**Comment by Centril** [comment]
+ `UNUSED_EXTERN_CRATES` is an idiom lint used when migrating to Rust 2018 (and activated by `#![deny(rust_2018_idioms)]`.
+ `ELIDED_LIFETIMES_IN_PATHS` is an idiom lint, e.g. on `fmt: Formatter`. Should become warn-by-default soon-ish?
+ `EXPLICIT_OUTLIVES_REQUIREMENTS` is an idiom lint, e.g. `struct Foo<T: 'a>(&'a T);` -- I'm not sure we want to promote this to a warning ever.
+ `MISSING_DOC_CODE_EXAMPLES` seems useful to keep as a way of enforcing documentation quality.
+ `UNSAFE_CODE` we should keep; this one is quite useful to `#![forbid(..)]` to ensure that you only have safe code in a crate. This could be done in clippy, but I think it's less common to run clippy as part of CI so having it in rustc is useful.
+ `MISSING_DOCS` is pretty useful to maintain docs quality; plenty of people use it.
+ `VARIANT_SIZE_DIFFERENCES` would fit in the perf category of clippy.

...other listed lints I'm less familiar with.

**Comment by Manishearth** [comment]
@centril the webpage isn't what makes clippy lints discoverable, clippy allow-by-default lints are a power user tool too, and we don't expect people to find them normally. The difference is that lint levels in clippy are typically one level higher than the level the lint would have in rustc; they're discovered because they trigger warnings.

I do think unsafe_code and the doc lints (also the idiom ones) should stay in rust, they are independently well-known. The rest could be moved around, probably.

**Comment by Manishearth** [comment]
Oh, also, the 1.0 RFC defined what goes in rustc and what doesn't: https://github.com/rust-lang/rfcs/pull/2476

It doesn't impose hard restrictions, but the rough idea was that anything that fits as a clippy `correctness` lint should probably be uplifted to rustc as a warn lint, and anything that doesn't (like most allow by default lints in rustc) probably should be in clippy.

**Comment by Centril** [comment]
@Manishearth 

> @Centril the webpage isn't what makes clippy lints discoverable, clippy allow-by-default lints are a power user tool too, and we don't expect people to find them normally. The difference is that lint levels in clippy are typically one level higher than the level the lint would have in rustc; they're discovered because they trigger warnings.

Let me rephrase... I think we can make rustc lints *more* discoverable and so more useful by having a webpage like that linked in a prominent place.

> Oh, also, the 1.0 RFC defined what goes in rustc and what doesn't: [rust-lang/rfcs#2476](https://github.com/rust-lang/rfcs/pull/2476)

(cc https://github.com/rust-lang/rust/issues/53224)

Sure; the RFC talks about *future* lints, not removing existing ones (because they already exist, and there's some churn for existing code-bases using them). I think if we want to downlift lints from rustc to clippy, let's go through them one by one and see which candidates there are and then the lang + compiler teams can confirm the deprecation.

**Comment by Manishearth** [comment]
I don't think that lint discoverability is that easy. The clippy lint list gets linked to from everything clippy outputs and people still don't know about it.

And yeah, the RFC talks about future lints, it also doesn't specify hard rules: I only brought it up because it specifies a standard that the community agreed upon, one that will be a useful default for making this judgement for existing lints if we are going down that path. It still should be case by case, I didn't say otherwise.

**Comment by Dylan-DPC** [comment]
ping from triage @Susurrus @oli-obk @Centril @Manishearth what's the update on this?

**Comment by Centril** [comment]
So to remove the existing lints, I'd like to see this first converted into `deny` + crater of that to see how widely this is in use. If the lints have some users, I think the churn of removing them isn't worth it.

**Comment by Susurrus** [comment]
I think there is a positive reception to adding missing trait lints for all the traits suggested by the API guidelines in Clippy based on feedback I've gotten to my PR there (rust-lang/rust-clippy#3752). That's where I've been spending my effort. However, that PR relies on the changes I propose here in `src/librustc/middle/lang_items.rs`. And there's the open question of what to do about the existing two lints here. As these aren't correctness lints, it's likely the right home for these are in Clippy long-term. One thing we could do now is add the lang item stuff that I need here and let Clippy add the lints as they see fit. Once they're in Clippy, we can reopen an issue about whether they belong in rustc too. Or do people think we need a consensus on our strategy now, as it doesn't make sense to add them to Clippy if they'd also be kept in the compiler.

**Comment by Centril** [comment]
I do think the existing missing_implementations lints are correctness lints in the sense that if they are used, the author considers it a bug if `Debug` isn't implemented. The standard library uses this to that effect iirc. I don't mind duplication in clippy for them tho; but if we are to remove them from here, there shouldn't be any notable amount of users.

**Comment by oli-obk** [comment]
> I do think the existing missing_implementations lints are correctness lints in the sense that if they are used, the author considers it a bug if Debug isn't implemented.

According to clippy's lint grouping rules, that's a restriction lint, not a correctness lint.

**Comment by Centril** [comment]
@oli-obk Yes fair enough.

**Comment by Dylan-DPC** [comment]
ping from triage @Sussurus @oli-obk any updates on this? are we blocked on anything?

**Comment by oli-obk** [comment]
@Susurrus 	I think as a first step we should do the refactorings suggested above for just the lints that are in rustc, and then implement the other lints in clippy.  This way we can punt the decision on deprecating the rustc lints into the future without being blocked or implementing all the lints in rustc

**Comment by Centril** [comment]
Removing the labels with that understanding :)

**Comment by Centril** [comment]
Ping from triage @Susurrus :)

**Comment by Dylan-DPC** [comment]
ping from triage @Susurrus Unfortunately we haven't heard from you on this in a while, so I'm closing the PR to keep things tidy. Don't worry though, if you'll have time again in the future please reopen this PR, we'll be happy to review it again! Thanks
